pkg/venafi/tpp/tpp.go

```diff
@@ -837,19 +839,42 @@ func (sp serverPolicy) toPolicy() (p endpoint.Policy) {
 		}
 		return escaped
 	}
-	const allAllowedRegex = ".*"
-	if len(sp.WhitelistedDomains) == 0 {
-		p.SubjectCNRegexes = []string{allAllowedRegex}
-	} else {
-		p.SubjectCNRegexes = make([]string, len(sp.WhitelistedDomains))
-		for i, d := range sp.WhitelistedDomains {
-			if sp.WildcardsAllowed {
-				p.SubjectCNRegexes[i] = addStartEnd(`([\p{L}\p{N}-*]+\.)*` + regexp.QuoteMeta(d))
-			} else {
-				p.SubjectCNRegexes[i] = addStartEnd(`([\p{L}\p{N}-]+\.)*` + regexp.QuoteMeta(d))
+	domainRegex := func(domain string, wildcardsAllowed bool) string {
+		requiresPrefix := false
+		if len(domain) > 0 && domain[0] == '.' {
+			domain = domain[1:]
+			requiresPrefix = true
+		}
+
+		switch {
+		case wildcardsAllowed && requiresPrefix:
+			return addStartEnd(`([\p{L}\p{N}-*]+\.)+` + regexp.QuoteMeta(domain))
+		case wildcardsAllowed && !requiresPrefix:
+			return addStartEnd(`([\p{L}\p{N}-*]+\.)*` + regexp.QuoteMeta(domain))
+		case !wildcardsAllowed && requiresPrefix:
+			return addStartEnd(`([\p{L}\p{N}-]+\.)+` + regexp.QuoteMeta(domain))
+		case !wildcardsAllowed && !requiresPrefix:
+			return addStartEnd(`([\p{L}\p{N}-]+\.)*` + regexp.QuoteMeta(domain))
+		}
+
+		panic("unreachable")
```

Comment:
I think I understand all these combinations.
And I can see that you're following the existing style by creating closure functions....although I can't see the point here, since all the variables needed are supplied as arguments.

And it's a shame we have to use regular expressions for this rather than parsing the domain name labels and comparing them right to left:
 * https://pkg.go.dev/github.com/miekg/dns#CompareDomainName

And it's a shame that the error message is "domain must match INCOMPREHENSIBLE_REGEXP rather than,
`domain must be a sub-domain of example.com`....

These rules seem treat the domain whitelist as always allowing sub-domains.
But what if I want to configure a TPP policy for the website team, to only allow certificates with the specific domains: `example.com` and `www.example.com` ?
I do not want to allow certificates that would clash with the mail server team who use domains `pop3.example.com` or `smtp.example.com`?

Reply:
`domainRegex` is called in two places, that is why it is a closure function.
I agree that regexes might not be ideal & TPP's domain name restrictions might not cover every use case, but fixing these issues is out-of-scope for this PR.

Reply:
Agreed. I wondered if the TPP web UI treats the domain whitelist items as suffixes and the documentation says that it  does

<img width="544" alt="image" src="https://github.com/jetstack/vcert/assets/978965/32aa3853-ed0e-47cc-838c-7cc28a350e28">


https://docs.venafi.com/Docs/current/TopNav/Content/Certificates/r-cert-object-settings-webadmin.php